### ISSUE TITLE
Question for the initialization of PairBiCRNet

### ISSUE THREAD

**Issue by ChicagoPark** [cs123951/bicardiac#1]
Thanks for the amazing work!

(1) I am trying to understand your code by running it, but because main.py does not exist, PairBiCRNet's input parameters are not clear to me. Can I ask how you set the input parameter below? Or can I ask, is it possible to ask the main.py?

class PairBiCRNet(GroupwiseModel):

    def __init__(self,
                 device_ids,
                 input_device,
                 output_device,
                 img_size,
                 down_out_channel_list,
                 num_layers,
                 hidden_dim_list,
                 kernel_size_list,
                 n_steps=0,
                 depthwise=False,
                 batch_first=True,
                 bias=True,
                 rnn_cell='ConvGRUCell',
                 cell_params=dict(),
                 dropout=-1,
                 conv_type='conv',
                 add_pair=False):
                 
(2) I also want to ask why the output displacement field shape is 3 (3@128x128x16), even though the input shape is 2 in (2@128x128x16). I understand input shape 2 is from complex and real channels, but I would like to ask output displacement field shape.

(3) From your notation of x, y, z, can I understand z as the number of phases?

Thank you so so much for your attention!

**Comment by cs123951** [comment]
Hello, Park, thank you for following my work.

**for Question (1)(3)**
 My work is for cine MR image which includes several 3D images in one sequence. 
As annotated in the code, the input tensor for the model is in the shape of { t* [batch, channel, x, y, z]}. 
- t is the length of the input sequence. The number of phases is t.
- batch can only be 1. 
- channel is 1 because the input image is gray image. 
- x, y, and z represent the dimension of images.  If you have 2d images, xyz becomes xy.

Since main.py is concerned with the configuration of dataset loader, it may not work with different dataset.
You can create a random tensor to test if the model worked:
```Python
model=PairBiCRNet(device_ids=[0],
             input_device=0,
             output_device0,
             img_size=[128,128,16],
             down_out_channel_list=[8,16],
             num_layers=1,
             hidden_dim_list=[16,16,16,16],
             kernel_size_list=[3,3,3,3]).cuda()
t = 20
inputs=torch.rand(t,1,1,128,128,16).cuda()
disp_tb3xyz, loss, loss_part=model(inputs)
```
You can refer to the following tutorial to start your own dataset training.
https://pytorch.org/tutorials/beginner/basics/quickstart_tutorial.html

**for Question (2)**
Suppose the input tensor is in the shape of { t* [batch, channel, x, y, z]}. Then we can form t-1 registration pairs. So the the output displacement field shape is (t-1, batch, 3, x, y, z). 
> I understand input shape 2 is from complex and real channels
the 2 in the input shape 2  may mean image pair which has two images (moving and fixing image).

Hope my answers can solve your problems. 
Sincerely,
Jiayi Lu

**Comment by ChicagoPark** [comment]
I truly appreciate your answer, Jiayi! It really helpful. I have one extra question. Sorry for your time.

(1) In my case, I have 2D images. I could see from your implementation that for 2D images, interp_mode is "bicubic." I believe this implementation is your intention!

(2) However, the output shape of the displacement field is still [9, 3, 320, 320] meaning 3d. Can I drop the last dimension and consider [9, 2, 320, 320] for the 2D processing? (e.g. [:, :2, ...])

(3) I am so sorry to ask many things, but can I ask how you used the displacement field to transform the image? Can I have the example code for that?

(4) From the paper, I understood you set the number of layers as 7, but your example is 1, and when I change it to 7, the code has an error. Can I ask what number of layers in implementation (same as your example, is it 1?)? In my case, based on your advice and paper, I tried below and can I ask is it what you did?

> regis_module = PairBiCRNet(device_ids=0, input_device=device, output_device=device, img_size=[320, 320], num_layers=7, kernel_size_list=[3, 3, 3, 3, 3, 3, 3], down_out_channel_list=[8, 16], hidden_dim_list=[16, 16, 16, 16, 16, 16, 16, 16]).cuda()

(5) I think it is important to do some normalization to apply the displacement field. Do you have any guidance on how to normalize the predicted displacement field?

I greatly appreciate your answer!! Thank you so much.

**Comment by cs123951** [comment]
Hello, 
Sorry for answering so late.

You questions (1)-(3) are all about  using the displacement field to transform the image. 
You can refer to https://github.com/voxelmorph/voxelmorph/blob/dev/voxelmorph/torch/layers.py to see how image is transformed. 
The core idea is `nnf.grid_sample(src, new_locs, align_corners=True, mode=self.mode)`

(4) I am wrong to write:
`regis_module = PairBiCRNet(device_ids=0, input_device=device, output_device=device, img_size=[320, 320], num_layers=7, kernel_size_list=[3, 3, 3, 3, 3, 3, 3], down_out_channel_list=[8, 16], hidden_dim_list=[16, 16, 16, 16, 16, 16, 16, 16]).cuda()`
Actually, num_layers=1, n_steps=7. 

(5) I do not understand what you mean. If you want to normalize the displacement field, just use a scale factor like what you have done to normalize images. Or do you mean regularization on the displacement field ? 

Hope my answers can solve your problems.
Sincerely,
Jiayi Lu

**Comment by ChicagoPark** [comment]
I found the implementation in another GitHub repository in yours! Thanks for the attention throughout the conversation!

def integrate_displacement(full_disp, grid, n_steps):
if n_steps > 0:
full_disp = full_disp / (2**n_steps)
for _ in range(n_steps):
full_disp = full_disp + grid_sample_without_grid(full_disp, full_disp, grid)
return full_disp

Thanks for the follow-up response, Jiayi! I greatly appreciated it.

Based on your advice to exactly mimic your paper's implementation, I was trying to add new argument n_steps=7. For the "n_steps=7", I could see the code requires "image_func.integrate_displacement". However, that function in GitHub does not include that function(integrate_displacement) in image_func.py.

I am pretty sure this is my last inquiry. Could you give me the implementation of "def integrate_displacement" in the image_func.py file?

Again, so sorry for the continuous question in your busy period.
